VotingApplication/VotingApplication.Web/Api/Controllers/PollVoteController.cs

```diff
@@ -1,93 +1,180 @@
-                    string lastPolledDate = queryMap["lastPoll"];
+                    ThrowError(HttpStatusCode.Forbidden, String.Format("Poll {0} has expired", pollId));
+                }
 
-                    if (lastPolledDate != null)
+
+                foreach (VoteRequestModel voteRequest in voteRequests)
+                {
+                    if (poll.Options.All(o => o.Id != voteRequest.OptionId))
                     {
-                        clientLastUpdated = UnixTimeToDateTime(long.Parse(lastPolledDate));
+                        ModelState.AddModelError("OptionId", "Option choice not valid for this poll");
                     }
+                }
+
+                // Poll specific validation
+                IVoteValidator voteValidator = _voteValidatorFactory.CreateValidator(poll.PollType);
+                voteValidator.Validate(voteRequests, poll, ModelState);
 
-                    if (poll.LastUpdated.CompareTo(clientLastUpdated) < 0)
+                if (!ModelState.IsValid)
+                {
+                    ThrowError(HttpStatusCode.BadRequest, ModelState);
+                }
+
+
+                Ballot ballot = poll
+                    .Ballots
+                    .SingleOrDefault(t => t.TokenGuid == tokenGuid);
+
+                if (ballot == null)
+                {
+                    if (poll.InviteOnly)
+                    {
+                        ThrowError(HttpStatusCode.Forbidden, String.Format("Token {0} not valid for this poll", tokenGuid));
+                    }
+                    else
                     {
-                        return new HttpResponseMessage(HttpStatusCode.NotModified);
+                        ballot = new Ballot() { TokenGuid = tokenGuid };
+                        poll.Ballots.Add(ballot);
                     }
                 }
 
-                var result = votes.Select(v => VoteToModel(v, poll)).ToList();
-                return Request.CreateResponse(HttpStatusCode.OK, result);
+                List<Vote> existingVotes = context
+                    .Votes
+                    .Include(v => v.Poll)
+                    .Where(v => v.Ballot.TokenGuid == tokenGuid && v.Poll.UUID == pollId)
+                    .ToList();
+
+                foreach (Vote contextVote in existingVotes)
+                {
+                    context.Votes.Remove(contextVote);
+                }
+
+                // For some reason, we don't have an addrange function on Entity Framework
+                foreach (VoteRequestModel voteRequest in voteRequests)
+                {
+                    Option option = context
+                        .Options
+                        .Single(o => o.Id == voteRequest.OptionId);
+
+                    Vote modelToVote = ModelToVote(voteRequest, ballot, option, poll);
+                    context.Votes.Add(modelToVote);
+
+                    // TODO: refactor the voteRequest model to be a ballotRequest instead. => only one voterName.
```

Comment:
Might have missed it, but is there a JIRA Issue for this? It's been a TODO for a while now.